_posts/2021-05-18-managing-react-application-state-with-context.md

```diff
@@ -0,0 +1,177 @@
+
+```javascript
+// countProvider.tsx
+type Action =
+  | { type: "increment" }
+  | { type: "decrement" }
+  | { type: "setValue", payload: { value: number } };
+type State = {
+    count: number,
+};
+type Dispatch = (action: Action) => void;
+```
+
+We follow this by setting up our *reducer* function which is exactly the same as it would be if we were using Redux; it is a pure function that takes the state, applies an action to it and returns the new state. I have set up handlers for each of the `Action`s I defined above which will allow us to `increment` and `decrement` the count as well as `setValue` of the count. Notice how I am careful not to directly change the existing state but instead copy the existing state with our changes applied on top of it.
+
+```javascript
+// countProvider.tsx
+function CountReducer(state: State, action: Action): State {
+  switch (action.type) {
+    case "increment":
+      return {
+        ...state,
+        count: state.count + 1,
+      };
+    case "decrement":
+      return {
+        ...state,
+        count: state.count - 1,
+      };
+    case "setValue": {
+      return {
+        ...state,
+        count: action.payload.value,
+      };
+    }
+    default:
+      throw Error(`Unhandled action type: ${action!.type}`);
+  }
+}
+```
+
+Next, I set up Context for both the `State` and `Dispatch` and provide a function component that determines how the `Reducer` we created above will be used. I also provide a way of setting the initial state which aids in testability since you can establish a given state in your tests before performing actions to see how the system behaves.
+
+```javascript
+// countProvider.tsx
+const CountStateContext = createContext<State | undefined>(undefined);
+const CountDispatchContext = createContext<Dispatch | undefined>(undefined);
+
+const CountProvider: FC<{ init?: Partial<State> }> = ({ children, init }) => {
+  const initialState: State = {
+    count: 0,
+    ...init,
+  };
+
+  const [state, dispatch] = useReducer(CountReducer, initialState);
+
+  return (
+    <CountStateContext.Provider value={state}>
+      <CountDispatchContext.Provider value={dispatch}>
+        {children}
+      </CountDispatchContext.Provider>
+    </CountStateContext.Provider>
+  );
+};
+```
+
+I then wrap our App in the `CountProvider` component, ensuring that all of the components that rely on the `State` or will perform `Action`s on that state are encapsulated as children of this component.
+
+```javascript
+// App.tsx
+function App() {
+  return (
+    <CountProvider>
+      <CountDisplay />
+      <CountButtons />
+      <CountInput />
+    </CountProvider>
```

Comment:
I know you've mentioned how this obliterates **prop-drilling**, I just hope in this example people know the Provider is available all the way through that react-tree-node!